### ISSUE TITLE
💅 `noBaseToString` ignores `toString(): string` declared on an ambient `declare class`, flagging every `.d.ts`-typed library value

### ISSUE THREAD

**Issue by viktor-horacek** [biomejs/biome#11320]
### Environment information

<details>

```bash
CLI:
  Version:                      2.5.8
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           windows

Environment:
  JS_RUNTIME_VERSION:           v24.19.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         bun/1.3.14

Biome Configuration:
  Status:                       Loaded successfully.
  Path:                         biome.json
  Formatter enabled:            false
  Linter enabled:               true

Linter:
  Enabled rules:
    nursery/noBaseToString
```

</details>

### Rule name

lint/nursery/noBaseToString

### Playground link

The rule belongs to the types domain and needs the scanner/type-inference engine, so the playground cannot run it. The reproduction below is two files with zero dependencies instead.

### Expected result

Only `index.ts:15` (`bad.toString()`) should be reported. Lines 12 and 13 resolve to a type that declares `toString(): string` and must not be flagged.

## Reproduction

Minimal, **zero dependencies**, two files.

`biome.json`:

```json
{
  "$schema": "https://biomejs.dev/schemas/2.5.8/schema.json",
  "formatter": { "enabled": false },
  "linter": {
    "enabled": true,
    "rules": {
      "preset": "none",
      "nursery": {
        "noBaseToString": "error"
      }
    }
  }
}
```

`index.ts` — `good` declares `toString(): string`, `bad` does not (the intended true positive):

```ts
declare class DeclaredDate {
  toString(): string;
}

declare class NoToString {
  value: number;
}

declare const good: DeclaredDate;
declare const bad: NoToString;

export const a = good.toString();
export const b = `date: ${good}`;

export const c = bad.toString();
```

```bash
bunx @biomejs/biome@2.5.8 check .
```

## Actual result

All three expressions are reported — including the two on `good`, whose type explicitly declares `toString(): string`:

```
index.ts:12:18 lint/nursery/noBaseToString ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × This expression will use the default object stringification format, such as [object Object], when stringified.

  > 12 │ export const a = good.toString();
       │                  ^^^^

index.ts:13:27 lint/nursery/noBaseToString ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × This expression will use the default object stringification format, such as [object Object], when stringified.

  > 13 │ export const b = `date: ${good}`;
       │                           ^^^^

index.ts:15:18 lint/nursery/noBaseToString ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × This expression will use the default object stringification format, such as [object Object], when stringified.

  > 15 │ export const c = bad.toString();
       │                  ^^^

Checked 2 files in 5ms. No fixes applied.
Found 3 errors.
```

`@typescript-eslint/no-base-to-string` on the byte-identical file reports **only line 15** (`typescript-eslint` 8.x, `projectService: true`; verified the rule was live via a plain-object control that it does report). The rule docs state "Same as `@typescript-eslint/no-base-to-string`", so this is a parity gap.

## Analysis

I narrowed which declaration shapes are affected — each linted in one run with only `noBaseToString` enabled:

| Declaration shape | Reported? |
| --- | --- |
| `declare class X { toString(): string }` | **yes — false positive** |
| `interface X { toString(): string }` | no (correct) |
| `type X = { toString(): string }` | no (correct) |
| `class X { toString(): string { … } }` (with body) | no (correct) |

So the miss is specific to **ambient class declarations** — class members without bodies. Interfaces, type literals, and concrete classes all resolve `toString` correctly.

This is why it is so visible in practice: every compiled library ships its classes as `declare class` in `.d.ts`, so any such value is flagged at every stringification site. Originally hit with `@internationalized/date` (used by React Aria / shadcn date components), where `CalendarDate`, `Time`, `CalendarDateTime` and `ZonedDateTime` each declare `toString(): string` (dist/types/src/CalendarDate.d.ts lines 46, 82, 137, 199) — both `date.toString()` and `` `${date}` `` are reported at every call site.

The `ignoredTypeNames` option does suppress it (`{ "options": { "ignoredTypeNames": ["CalendarDate"] } }`), but that requires enumerating every class of every dependency, so it is a workaround rather than a fix.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
Please use the playground 


.https://biomejs.dev/playground/?lintRules=useReactCompiler&tab=settings&pane=Diagnostics

**Comment by biome-cookie** [comment]
Hello @viktor-horacek, please provide a minimal reproduction. You can use one of the following options:

- Provide a link to [our playground](https://biomejs.dev/playground), if it's applicable.
- Provide a link to GitHub repository. To easily create a reproduction, you can use our interactive CLI via `npm create @biomejs/biome-reproduction`

Issues marked with `S-Needs repro` will be **closed** if they have **no activity within 3 days**.

**Comment by biome-cookie** [comment]
This issue has been automatically closed because it has been inactive for more than 3 days and it does not have a minimal reproduction.
